lib/Pool.js

```diff
@@ -109,8 +109,11 @@ Pool.prototype.releaseConnection = function releaseConnection(connection) {
     }
 
     if (connection._purge) {
-      // purge connection from pool
-      this._purgeConnection(connection);
+      // restore the switched user
+      this.restoreUser(connection, function (err) {
+        if (err) return connection.destroy();
```

Comment:
Why call `connection.destroy();` on error instead of `this._purgeConnection(connection);` ?